release-2.0: storage/engine: limit the size of batch groups

petermattis opened the pull request:
Backport 1/1 commits from #27895.

/cc @cockroachdb/release

Queuing this up so we don't forget about it.

---

Limit the size of batch groups to 1 MB. A larger batch can still be
committed, but it won't be grouped with any other batches.

Fixes #27865

Release note (bug fix): Limit the size of "batch groups" when committing
a batch to RocksDB to avoid rare scenarios in which multi-gigabyte batch
groups are created which can cause a server to run out of memory when
replaying the RocksDB log at startup.

petermattis committed:
storage/engine: limit the size of batch groups

Limit the size of batch groups to 1 MB. A larger batch can still be
committed, but it won't be grouped with any other batches.

Fixes #27865

Release note (bug fix): Limit the size of "batch groups" when committing
a batch to RocksDB to avoid rare scenarios in which multi-gigabyte batch
groups are created which can cause a server to run out of memory when
replaying the RocksDB log at startup.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28009)
<!-- Reviewable:end -->

petermattis commented:
Ping. I think this should be merged in anticipation of 2.0.5.

benesch reviewed (approved):
Backport LGTM, but deferring to @bdarnell on the risk evaluation.

petermattis commented:
@benesch Do you know what is going on with the teamcity failure:

```
[16:31:59][Build archive] set -euo pipefail; \
[16:31:59][Build archive]   ret=$(cd ./pkg/sql/parser/gen && goyacc -p sql -o sql.go.tmp sql.y); \
[16:31:59][Build archive]   if expr "$ret" : ".*conflicts" >/dev/null; then \
[16:31:59][Build archive]     echo "$ret"; exit 1; \
[16:31:59][Build archive]   fi
[16:31:59][Build archive] (echo "// Code generated by make. DO NOT EDIT."; \
[16:31:59][Build archive]  echo "// GENERATED FILE DO NOT EDIT"; \
[16:31:59][Build archive]  echo; \
[16:31:59][Build archive]  echo "package lex"; \
[16:31:59][Build archive]  echo; \
[16:31:59][Build archive]  grep '^const [A-Z][_A-Z0-9]* ' pkg/sql/parser/gen/sql.go.tmp) > pkg/sql/lex/tokens.go
[16:31:59][Build archive] (echo "// Code generated by goyacc. DO NOT EDIT."; \
[16:31:59][Build archive]  echo "// GENERATED FILE DO NOT EDIT"; \
[16:31:59][Build archive]  cat pkg/sql/parser/gen/sql.go.tmp | \
[16:31:59][Build archive]  sed -E 's/^const ([A-Z][_A-Z0-9]*) =.*$/const \1 = lex.\1/g') > pkg/sql/parser/sql.go
[16:32:05][Build archive] build/node-run.sh -C ./pkg/ui yarn install
[16:32:05][Build archive] error An unexpected error occurred: "https://registry.yarnpkg.com/svgo/-/svgo-0.7.2.tgz: Request failed \"500 Internal Server Error\"".
[16:32:05][Build archive] yarn install v1.5.1
[16:32:05][Build archive] make: *** [pkg/ui/yarn.installed] Error 1
[16:32:05][Build archive] [1/5] Validating package.json...
[16:32:05][Build archive] [2/5] Resolving packages...
[16:32:05][Build archive] [3/5] Fetching packages...
[16:32:05][Build archive] info If you think this is a bug, please open a bug report with the information provided in "/go/src/github.com/cockroachdb/cockroach/pkg/ui/yarn-error.log".
[16:32:05][Build archive] info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
[16:32:05][Build archive] Makefile:322: recipe for target 'pkg/ui/yarn.installed' failed
[16:32:06][Build archive] Process exited with code 2
```

benesch commented:
Looks like the Yarn registry is down. 🙄

Maybe just retry the build in a bit? The fix to avoid hitting the Yarn registry on every build is probably too large to backport.

petermattis commented:
> Maybe just retry the build in a bit? 

I can do that.

petermattis commented:
bors r=bdarnell,benesch

petermattis commented:
@tschottdorf (or anyone) is this CI failure known (this is the release-2.0 branch):

```
[/runMode=local] test_log_scope.go:81: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/logTestDecommission_runMode=local206441087
test_log_scope.go:62: use -show-logs to present logs inline
decommission_test.go:338: csv input:
	id,is_live,gossiped_replicas,is_decommissioning,is_draining
	1,true,0,true,false
	
	All target nodes report that they hold no more data. Please verify cluster health before removing the nodes.
	
	expected:
	[][]string{
	    {"id", "is_live", "gossiped_replicas", "is_decommissioning", "is_draining"},
	    {"1", "true", "0", "true", "true"},
	    {"All target nodes report that they hold no more data. Please verify cluster health before removing the nodes."},
	}
	errors:<nil>
	row #2, col #5: found "false" which does not match "true"
```

petermattis committed:
storage: fix batch commit performance regression

PR #27895 had a subtle bug. Only the leader at the head of the pending
queue was waiting for the previous commit to finish. The other leaders
were marching right past the wait loop and committing their singleton
batch. This was not a correctness problem, only a performance
problem. But fixing it did reveal two other correctness problems. First,
we need to broadcast to wake all waiters on the cond. Prior to fix the
first bug there was only one cond waiter, so this was only a lurking
problem. Second, we need to cap the capacity of the slice passed to the
syncer to prevent concurrent mutation (this showed up readily as a race
condition).

Release note (performance improvement): Fix batch commit performance
regression that reduced write performance by 20%.

petermattis commented:
Updated this PR to include the performance fix as well.

petermattis commented:
bors r=benesch
